File: cpp/horus/pb/detection_merger/service_handler.cpp
#include "horus/pb/detection_merger/service_handler.h"

#include <utility>

#include "horus/future/resolved.h"

namespace horus {
namespace sdk {
namespace pb {

AnyFuture<pb::RpcMessage> DetectionMergerServiceHandler::Handle(const RpcContext& context,
                                                                pb::RpcMessage&& request) noexcept(false) {
  const pb::RpcMessage owned_request{std::move(request)};
  switch (owned_request.method_id()) {
    case 1: {
      return DoDispatch(&DetectionMergerServiceHandler::Subscribe, context, owned_request);
    }
    case 2: {
      return DoDispatch(&DetectionMergerServiceHandler::Unsubscribe, context, owned_request);
    }
    default: {
      throw horus_internal::UnknownRpcMethodError{};
    }
  }
}

AnyFuture<pb::DefaultSubscribeResponse> DetectionMergerServiceHandler::Subscribe(const RpcContext& context,
                                                                                 pb::DefaultSubscribeRequest&& request) noexcept(false) {
  static_cast<void>(pb::DefaultSubscribeRequest{std::move(request)});
  return ResolveWith(DetectionMergerSubscriberServiceSubscribers().Add<pb::DefaultSubscribeResponse>(context));
}

AnyFuture<pb::DefaultUnsubscribeResponse> DetectionMergerServiceHandler::Unsubscribe(const RpcContext& context,
                                                                                     pb::DefaultUnsubscribeRequest&& request) noexcept(false) {
  static_cast<void>(pb::DefaultUnsubscribeRequest{std::move(request)});
  return ResolveWith(DetectionMergerSubscriberServiceSubscribers().Remove<pb::DefaultUnsubscribeResponse>(context));
}

AnyFuture<pb::RpcMessage> DetectionMergerSubscriberServiceHandler::Handle(const RpcContext& context,
                                                                          pb::RpcMessage&& request) noexcept(false) {
  const pb::RpcMessage owned_request{std::move(request)};
  switch (owned_request.method_id()) {
    case 1: {
      return DoDispatch(&DetectionMergerSubscriberServiceHandler::BroadcastDetection, context, owned_request);
    }
    default: {
      throw horus_internal::UnknownRpcMethodError{};
    }
  }
}

}  // namespace pb
}  // namespace sdk
}  // namespace horus

File: py/horus/pb/detection_merger/service_handler.py
import abc as _abc
from ...rpc.base_handler import BaseRpcHandler, RpcMessage, WebSocket

from ..detection_service.detection_pb2 import DetectionEvent
from ..rpc_pb2 import DefaultSubscribeRequest, DefaultSubscribeResponse, DefaultUnsubscribeRequest, DefaultUnsubscribeResponse

class DetectionMergerServiceHandler(BaseRpcHandler, metaclass=_abc.ABCMeta):
    """
    Service for merging detection results from multiple sources.
    """

    service_full_name = "horus.pb.DetectionMergerService"
    service_id = 11

    def __init__(self, ws: WebSocket):
        super().__init__(ws, {
            1: (DefaultSubscribeRequest, DefaultSubscribeResponse, self.subscribe),
            2: (DefaultUnsubscribeRequest, DefaultUnsubscribeResponse, self.unsubscribe),
        })

    @_abc.abstractmethod
    async def subscribe(self, request: DefaultSubscribeRequest) -> DefaultSubscribeResponse:
        """
        Subscribe to receive merged detection results.

        This function has a default implementation in C++, but not in Python.
        """
        raise NotImplementedError

    @_abc.abstractmethod
    async def unsubscribe(self, request: DefaultUnsubscribeRequest) -> DefaultUnsubscribeResponse:
        """
        Unsubscribe from merged detection results.

        This function has a default implementation in C++, but not in Python.
        """
        raise NotImplementedError

class DetectionMergerSubscriberServiceHandler(BaseRpcHandler, metaclass=_abc.ABCMeta):
    """
    Service for receiving merged detection results.
    """

    service_full_name = "horus.pb.DetectionMergerSubscriberService"
    service_id = 16

    def __init__(self, ws: WebSocket):
        super().__init__(ws, {
            1: (DetectionEvent, None, self.broadcast_detection),
        })

    @_abc.abstractmethod
    async def broadcast_detection(self, request: DetectionEvent) -> None:
        """
        Receive merged detection results.
        """
        raise NotImplementedError
